src/mds/schedule/copySetScheduler.cpp

```diff
@@ -38,27 +38,84 @@ int CopySetScheduler::Schedule() {
-    // 1. collect the chunkserver list and copyset list of the cluster, then
-    //    collect copyset on every online chunkserver
-    auto copysetList = topo_->GetCopySetInfosInLogicalPool(lid);
-    auto chunkserverList = topo_->GetChunkServersInLogicalPool(lid);
-    std::map<ChunkServerIdType, std::vector<CopySetInfo>> distribute;
-    SchedulerHelper::CopySetDistributionInOnlineChunkServer(
-        copysetList, chunkserverList, &distribute);
-    if (distribute.empty()) {
-        LOG(WARNING) << "no not-retired chunkserver in topology";
-        return UNINTIALIZE_ID;
+int CopySetScheduler::PenddingCopySetSchedule(const std::map<ChunkServerIdType,
+                                    std::vector<CopySetInfo>> &distribute) {
+    int oneRoundGenOp = 0;
+    // for every chunkserver, find one copy set to migrate out
+    for (auto it = distribute.begin(); it != distribute.end(); it++) {
+        ChunkServerIdType source = it->first;
+        int copysetNum = it->second.size();
+        if (copysetNum == 0) {
+            continue;
+        }
+
+        // find one copy set to migrate out from source chunkserver
+        for (auto info : it->second) {
+            // does not meet the basic conditions
+            if (!CopySetSatisfiyBasicMigrationCond(info)) {
+                continue;
+            }
+
+            auto target = SelectBestPlacementChunkServer(info, source);
+            if (target == UNINTIALIZE_ID) {
+                LOG(WARNING) << "copysetScheduler can not select chunkServer "
+                                "to migrate " << info.CopySetInfoStr()
+                            << ", which replica: " << source << " is pendding";
+                continue;
+            }
+
+            Operator op = operatorFactory.CreateChangePeerOperator(
+                info, source, target, OperatorPriority::HighPriority);
+            op.timeLimit = std::chrono::seconds(changeTimeSec_);
+
+            // add operator
+            if (!opController_->AddOperator(op)) {
+                LOG(INFO) << "copysetSchduler add op " << op.OpToString()
+                        << " fail, copyset has already has operator";
+                continue;
+            }
+
+            // create copyset
+            if (!topo_->CreateCopySetAtChunkServer(info.id, target)) {
+                LOG(ERROR) << "pendding copyset scheduler create "
+                        << info.CopySetInfoStr()
+                        << " on chunkServer: " << target
+                        << " error, delete operator" << op.OpToString();
+                opController_->RemoveOperator(info.id);
+                continue;
+            }
+
+            oneRoundGenOp++;
+
+            LOG(INFO) << "copysetScheduler create " << info.CopySetInfoStr()
+                    << "on chunkserver:" << target
+                    << " success. generator op: "
+                    << op.OpToString() << "success";
+            break;
+        }
```

Comment:
Can I know if the migration is complete?

Reply:
The function int StatusTool::ChunkServerListCmd() in status_tool.cpp  is modified. It can print the pendding status of chunkserver.